### PR TITLE
Version 9 - reassembling the tractor

### PR DESCRIPTION
**Pull request by jendakol**
This branch contains many changes at once - unfortunately. The API changes are described in the readme and migration guide so go see it for actual content of this huge PR.

### CODE REVIEW

**Comment by jendakol** [comment]
@augi As this version has been production tested for a few weeks, I guess we can merge it and release as RC :-) Don't make much trouble yourself with reading the whole code... it's really unpleasant experience as the whole library was disassembled and assembled again - while API stays mostly without changes, internally it's a bit different (and better, I believe... it was a pain to maintain from time to time).

**Comment by augi** [comment]
I've read the migration and it looks good to me: https://github.com/avast/rabbitmq-scala-client/blob/PreparingV9/Migration-8-9.md

Also would be nice to hear from @mi-char if it is OK for him.

**Comment by augi** [comment]
@jendakol What was the main driver, to make this huge refactoring?

**Review by augi — APPROVED** [review]

**Review by mi-char — COMMENTED** [review]
I've checked mainly the API changes and agree with them with some exceptions.

I'm not sure regarding the CorrelationID behavior that seems to be inconsistent, on some places it behaves like a mandatory value, but for instance, in the `DeliveryMetadata` case class it works with the `-none-` value that represents the empty/missing value. Almost the same situation is with the `MessageId` property.

I'm also not sure if it is wise to heavily depend on the Blocker class, that was discontinued in the Cats Effect 3.

**Comment by jendakol** [comment]
@augi Well, there had to be _some_ refactoring done because I realized there is a big problem with the poisoned message handler which was unable to catch timed-out messages (by design 😑). While solving that, I had to do some breaking changes and when I hit some "suboptimal places" (some stuff that was not solved ideally in the past, either because of lack of wisdom 😄 or to preserve the compatibility) I decided to change it everything. Unfortunately, as that big refactoring was not planned, I was unable to split the change into some meaningful commits which is a shame but... 🤷🏻‍♂️

File: api/src/main/scala/com/avast/clients/rabbitmq/api/DeliveryResult.scala [hunk 1]
```diff
@@ -14,6 +14,6 @@ object DeliveryResult {
   case object Retry extends DeliveryResult
 
   /** The message cannot be processed but is worth - it will be requeued to the bottom of the queue. */
-  case class Republish(newHeaders: Map[String, AnyRef] = Map.empty) extends DeliveryResult
+  case class Republish(isPoisoned: Boolean = true, newHeaders: Map[String, AnyRef] = Map.empty) extends DeliveryResult
```
[the remark this answers is not in the document]
Comment (jendakol) [reply api/src/main/scala/com/avast/clients/rabbitmq/api/DeliveryResult.scala:17]: The idea behind this is that you may want to republish the delivery for some reason which is out of your hands - e.g. you call a 3rd party service that returns HTTP 503. It's unlikely that is a problem with the specific delivery so it shouldn't be thrown away, still you want to _republish_ it so it doesn't block processing of some other delivery which might not need the failing dependency.

File: README.md [hunk 2]
```diff
@@ -75,81 +81,82 @@ val blockingExecutor: ExecutorService = ???
...
 
 It seems quite natural to process RabbitMQ queue with a streaming app.
-[`StreamingRabbitMQConsumer`](core/src/main/scala/com/avast/clients/rabbitmq/StreamingRabbitMQConsumer.scala) provides you an
+[`StreamingRabbitMQConsumer`](api/src/main/scala/com/avast/clients/rabbitmq/api/RabbitMQStreamingConsumer.scala) provides you an
 [`fs2.Stream`](https://fs2.io/) through which you can easily process incoming messages in a streaming way.
 
-Notice: Using this functionality requires you to know some basics of [FS2](https://fs2.io/guide.html#overview) library. Please see it's official
-guide if you're not familiar with it first.
+Notice: Using this functionality requires you to know some basics of [FS2](https://fs2.io/guide.html#overview) library. Please see it's
+official guide if you're not familiar with it first.
```
[the remark this answers is not in the document]
Comment (jendakol) [reply README.md:136]: Oh. That is unfortunate 😑 What I should rename this to then? 🤔

[the remark this answers is not in the document]
Comment (mi-char) [reply README.md:136]: I don't have a better idea than the name `StreamingPushRabbitMQConsumer` + renaming the Readme chapter to `Streaming Push consumer` + explicit notice that it is not an implementation of the RMQ Streams consumer.

[the remark this answers is not in the document]
Comment (jendakol) [reply README.md:136]: I consider `Push` to be a misleading name for everyone who have no idea that sth like RMQ Streams exists 😐

**Comment by jendakol** [comment]
OK guys @mi-char @augi , I've pushed the changes we spoke about in person. See the latest commits and let me know what you think about that.   
I didn't implement the hierarchical CorrelationId strategy, however, it's ready to be implemented later or anyone can implement his own. (The main reason is I didn't find anything much relevant about that.)  
Also, the strategy doesn't have default value so everyone has to manually choose which strategy to use... but I'm not sure if it's a better solution than to keep it like it was, with the "FromPropertiesOrRandomNew".

**Review by mi-char — APPROVED** [review]
I must say that the latest changes perfectly fit into the proposed API.

~~I'm still not sure about the `DeliveryResult.Republish(isPoisoned: Boolean = true` as a default value.~~
I'm still not sure about the naming of the `DeliveryResult.Republish.isPoisoned` property as it is too vague.

Regarding the default value for the `CorrelationIdStrategy`, I'm for using `FromPropertiesOrRandomNew` or introducing `CorrelationIdStrategy.NoOp` (that would be more backward compatible) as a default.

Good job 👍

**Pull request merged**